### PR TITLE
feat(SPV-1056): add handling paymail output in new transaction drafts

### PR DESCRIPTION
**Pull request by dorzepowski**
⚠️ This change doesn't handle the sender paymail yet, but I want the PR to be easier to read, because adding sender will add changes in additional few files.

# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: engine/paymail/paymail_service_client.go [hunk 1]
```diff
@@ -73,6 +74,35 @@ func (s *service) GetCapabilities(ctx context.Context, domain string) (*paymail.
 	return &response.CapabilitiesPayload, nil
 }
 
+// GetP2PDestinations will ask a paymail host on given address for P2P destinations.
+func (s *service) GetP2PDestinations(ctx context.Context, address *paymail.SanitisedPaymail, satoshis uint) (*paymail.PaymentDestinationPayload, error) {
```
Comment (chris-4chain) [inline engine/paymail/paymail_service_client.go:78]: In our project - I can see - that the variables types for `satoshis` are inconsistent. (uint, uint64)
I know that on 64-but machines there is no difference between them. 
But... 
Since we are now planning to fix things in spv-wallet - I think we could establish ONE type for these.

File: engine/paymail/paymail_service_client.go [hunk 2]
```diff
@@ -184,3 +217,14 @@ func (s *service) extractP2P(capabilities *paymail.CapabilitiesPayload) (success
 	}
 	return
 }
+
+func (s *service) validatePaymentDestinationResponse(response *paymail.PaymentDestinationResponse, satoshis uint) error {
+	var sum uint64
+	for _, out := range response.Outputs {
+		sum += out.Satoshis
+	}
+	if sum != uint64(satoshis) {
```
Comment (chris-4chain) [inline engine/paymail/paymail_service_client.go:226]: I think, if we design it better - the conversion will not be necessary.

File: engine/tester/paymailmock/mock_paymail_host_responses.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+// MockedP2PDestinationResponse is a mocked response for the P2P destinations endpoint
+type MockedP2PDestinationResponse struct {
+	Reference string
+	Scripts   []string
+	Satoshis  []uint
+}
+
+var _mockedLockingScripts = []string{
```
Comment (chris-4chain) [inline engine/tester/paymailmock/mock_paymail_host_responses.go:17]: is the underscore necessary - what it indicates in this context?

File: engine/tester/paymailmock/mock_paymail_host_responses.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	return obj{
+		"outputs":   outs,
+		"reference": m.Reference,
+	}
+}
+
+// P2PDestinationsForSats returns a mocked response for the P2P destinations endpoint
+func P2PDestinationsForSats(satoshis uint, moreSatoshis ...uint) *MockedP2PDestinationResponse {
+	outputSats := append([]uint{satoshis}, moreSatoshis...)
```
Comment (chris-4chain) [inline engine/tester/paymailmock/mock_paymail_host_responses.go:50]: nice 💯 😎

File: engine/transaction/draft/create_paymail_draft_test.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+			},
+			expectedError: "paymail host is responding with error",
+		},
+		"paymail host p2p destinations is returning not found": {
+			paymailHostScenario: func(paymailClient *paymailmock.PaymailClientMock) {
+				paymailClient.
+					WillRespondWithP2PCapabilities().
+					WillRespondOnCapability(paymail.BRFCP2PPaymentDestination).
+					WithNotFound()
+			},
```
Comment (chris-4chain) [inline engine/transaction/draft/create_paymail_draft_test.go:226]: The `expectedError` field is missing here. Did you do it on purpose?

File: engine/transaction/draft/create_paymail_draft_test.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+		// then:
+		require.NoError(t, err)
+		require.NotNil(t, draftTx)
+
+		// and:
+		annotations := draftTx.Annotations
+		require.Len(t, annotations.Outputs, 2)
+		assert.Equal(t, transaction.BucketBSV, annotations.Outputs[0].Bucket)
+		assert.Equal(t, transaction.BucketBSV, annotations.Outputs[1].Bucket)
+		// TODO: add assertions for paymail annotations
```
Comment (chris-4chain) [inline engine/transaction/draft/create_paymail_draft_test.go:115]: Is it a leftover or you're gonna do it in te next task?

File: engine/transaction/draft/create_paymail_draft_test.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+	"github.com/bitcoin-sv/spv-wallet/engine/tester"
+	"github.com/bitcoin-sv/spv-wallet/engine/tester/paymailmock"
+	"github.com/bitcoin-sv/spv-wallet/engine/transaction"
+	"github.com/bitcoin-sv/spv-wallet/engine/transaction/draft"
+	"github.com/bitcoin-sv/spv-wallet/engine/transaction/draft/outputs"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCreatePaymailDraft(t *testing.T) {
```
Comment (chris-4chain) [inline engine/transaction/draft/create_paymail_draft_test.go:18]: Will it be beneficial to test some edge cases:
- transaction with zero satoshis
- transaction with extremely large values

File: engine/transaction/draft/draft_service.go [hunk 8]
```diff
@@ -33,10 +34,14 @@ func (s *service) Create(ctx context.Context, spec *TransactionSpec) (*Transacti
 	if spec == nil {
 		return nil, txerrors.ErrDraftSpecificationRequired
 	}
-	outputs, annotations, err := spec.outputs(ctx)
+
+	c := evaluation.NewContext(ctx, s.logger, s.paymailService)
```
Comment (chris-4chain) [inline engine/transaction/draft/draft_service.go:38]: Quite smart solution with the `context` 😄

File: models/request/draft_transaction_json.go [hunk 9]
```diff
@@ -36,6 +43,14 @@ func expandOutputForMarshaling(output Output) (any, error) {
 			Type:   o.GetType(),
 			Output: &o,
 		}, nil
+	case paymailreq.Output:
```
Comment (chris-4chain) [inline models/request/draft_transaction_json.go:46]: Will you consider my suggestion about this part of code I gave you some time ago?
- Or you're fine with this solution and "if it works - don't touch it?"

File: models/request/optional/optional.go [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package optional
+
+type Param[T any] *T
+
+func Of[T any](v T) Param[T] {
+	return &v
+}
```
Comment (chris-4chain) [inline models/request/optional/optional.go:7]: Maybe such functionality we could move to more "general" package somewhere closer to the "root"?

File: models/request/paymail/paymail_output.go [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package paymail
+
+import (
+	"github.com/bitcoin-sv/spv-wallet/models/request/optional"
+)
+
+type Output struct {
+	To       string                 `json:"to"`
+	Satoshis uint                   `json:"satoshis"`
+	From     optional.Param[string] `json:"from,omitempty"`
```
Comment (chris-4chain) [inline models/request/paymail/paymail_output.go:10]: 💯

**Review by wregulski — APPROVED** [review]

File: models/request/optional/optional.go (hunk 10 above)
Comment (dorzepowski) [reply models/request/optional/optional.go:7]: ok, but still I would like to keep it in a models package, to make it sharable,

WDYT?

Comment (chris-4chain) [reply models/request/optional/optional.go:7]: Good idea 👍

File: engine/transaction/draft/create_paymail_draft_test.go (hunk 7 above)
Comment (dorzepowski) [reply engine/transaction/draft/create_paymail_draft_test.go:18]: yes it will, thank you

File: engine/transaction/draft/create_paymail_draft_test.go (hunk 6 above)
Comment (dorzepowski) [reply engine/transaction/draft/create_paymail_draft_test.go:115]: leftover I don't know why I read it previously as add assertions for sender 🙈

File: engine/tester/paymailmock/mock_paymail_host_responses.go (hunk 3 above)
Comment (dorzepowski) [reply engine/tester/paymailmock/mock_paymail_host_responses.go:17]: no I start with some const approach and it's leftover 🙈

File: engine/transaction/draft/create_paymail_draft_test.go (hunk 5 above)
Comment (dorzepowski) [reply engine/transaction/draft/create_paymail_draft_test.go:226]: No, my mistake

File: models/request/optional/optional.go (hunk 10 above)
Comment (dorzepowski) [reply models/request/optional/optional.go]: Moved

File: engine/transaction/draft/create_paymail_draft_test.go (hunk 7 above)
Comment (dorzepowski) [reply engine/transaction/draft/create_paymail_draft_test.go:19]: added tests for zero and negative satoshis and also for too large amount on JSON side

File: engine/transaction/draft/create_paymail_draft_test.go (hunk 6 above)
Comment (dorzepowski) [reply engine/transaction/draft/create_paymail_draft_test.go]: actually I forgot to add those assertions, now they are added.

File: engine/transaction/draft/create_paymail_draft_test.go (hunk 5 above)
Comment (dorzepowski) [reply engine/transaction/draft/create_paymail_draft_test.go:259]: fixed

File: engine/tester/paymailmock/mock_paymail_host_responses.go (hunk 3 above)
Comment (dorzepowski) [reply engine/tester/paymailmock/mock_paymail_host_responses.go]: changed

File: engine/paymail/paymail_service_client.go (hunk 2 above)
Comment (dorzepowski) [reply engine/paymail/paymail_service_client.go]: I'm using now bsv.Satoshis from models package

File: engine/paymail/paymail_service_client.go (hunk 1 above)
Comment (dorzepowski) [reply engine/paymail/paymail_service_client.go]: I'm using now bsv.Satoshis from models package

**Review by chris-4chain — APPROVED** [review]
